x/dao/keeper/delegation.go

```diff
@@ -0,0 +1,157 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/onomyprotocol/onomy/x/dao/types"
+)
+
+// ReBalanceDelegation re-balances the DAO staking among validators bases on the current validators self bond.
+func (k Keeper) ReBalanceDelegation(ctx sdk.Context) error {
+	vals := k.stakingKeeper.GetAllValidators(ctx)
+	targetDelegations := k.getTargetDelegationState(ctx, vals)
+	return k.reBalanceDaoStaking(ctx, vals, targetDelegations)
+}
+
+// GetDaoDelegationSupply returns total amount of the treasury bonded coins.
+func (k Keeper) GetDaoDelegationSupply(ctx sdk.Context) sdk.Dec {
+	return k.getDaoDelegationSupply(ctx)
+}
+
+// getTargetDelegationState builds a map of the validators and the stake amount they should have now.
+// if the validator is not in the map, the DAO stake is zero.
+func (k Keeper) getTargetDelegationState(ctx sdk.Context, vals []stakingtypes.Validator) map[string]sdk.Dec {
+	maxStakingCommissionRate := k.StakingMaxCommissionRate(ctx)
+	valsSelfBonds := make(map[string]sdk.Dec) // the key is OperatorAddress
+	valsSelfBondsSupply := sdk.ZeroDec()
+	for _, val := range vals {
+		if !val.IsBonded() {
+			continue
+		}
+		if val.GetCommission().GT(maxStakingCommissionRate) {
+			continue
+		}
+
+		valOper := val.GetOperator()
+		valAddr := sdk.AccAddress(valOper)
+		selfDelegation, found := k.stakingKeeper.GetDelegation(ctx, valAddr, valOper)
+		if !found || selfDelegation.GetShares().IsZero() {
+			continue
+		}
+		selfDelegationAmount := val.TokensFromShares(selfDelegation.GetShares())
+		valsSelfBonds[valOper.String()] = selfDelegationAmount
+		valsSelfBondsSupply = valsSelfBondsSupply.Add(selfDelegationAmount)
+	}
+
+	daoDelegationSupply := k.getDaoDelegationSupply(ctx)
+	daoBondDenomSupply := k.treasuryBondDenomAmount(ctx).ToDec().Add(daoDelegationSupply)
+
+	daoBondDenomToDelegate := daoBondDenomSupply.Sub(daoBondDenomSupply.Mul(k.StakingTokenPoolRate(ctx)))
+
+	targetDelegationState := make(map[string]sdk.Dec) // the key is OperatorAddress
+	for valAddr, selfDelegationAmt := range valsSelfBonds {
+		valDaoDelegationAmt := selfDelegationAmt.Quo(valsSelfBondsSupply).Mul(daoBondDenomToDelegate)
```

Comment:
Another `Quo` and `Mul` reversal should happen here

Reply:
Agree, updated.